core/signer/signer.go

```diff
@@ -63,40 +64,80 @@ func init() {
 
 // Signer processes mutations and sends them to the trillian map.
 type Signer struct {
-	realm     string
 	mapID     int64
 	tmap      trillian.TrillianMapClient
 	logID     int64
-	sths      appender.Remote
+	tlog      trillian.TrillianLogClient
 	mutator   mutator.Mutator
 	mutations mutator.Mutation
 	factory   transaction.Factory
 }
 
 // New creates a new instance of the signer.
-func New(realm string,
-	mapID int64,
+func New(mapID int64,
 	tmap trillian.TrillianMapClient,
 	logID int64,
-	sths appender.Remote,
+	tlog trillian.TrillianLogClient,
 	mutator mutator.Mutator,
 	mutations mutator.Mutation,
 	factory transaction.Factory) *Signer {
 	return &Signer{
-		realm:     realm,
 		mapID:     mapID,
 		tmap:      tmap,
 		logID:     logID,
-		sths:      sths,
+		tlog:      tlog,
 		mutator:   mutator,
 		mutations: mutations,
 		factory:   factory,
 	}
 }
 
+// Initialize inserts the object hash of an empty struct into the log if it is empty.
+// This keeps the log leaves in-sync with the map which starts off with an
+// empty log root at map revision 0.
+func (s *Signer) Initialize(ctx context.Context) error {
+	logRoot, err := s.tlog.GetLatestSignedLogRoot(ctx, &trillian.GetLatestSignedLogRootRequest{
+		LogId: s.logID,
+	})
+	if err != nil {
+		return fmt.Errorf("GetLatestSignedLogRoot(%v): %v", s.logID, err)
+	}
+	mapRoot, err := s.tmap.GetSignedMapRoot(ctx, &trillian.GetSignedMapRootRequest{
+		MapId: s.mapID,
+	})
+	if err != nil {
+		return fmt.Errorf("GetSignedMapRoot(%v): %v", s.mapID, err)
+	}
+
+	// If the tree is empty and the map is empty,
+	// add the empty map root to the log.
+	if logRoot.GetSignedLogRoot().GetTreeSize() == 0 &&
+		mapRoot.GetMapRoot().GetMapRevision() == 0 {
+		smrJSON, err := json.Marshal(mapRoot.GetMapRoot())
+		idHash := sha256.Sum256(smrJSON)
```

Comment:
The log server sets the leaf hash for us. It's only duplicate work on our end to do it here.